lib/testing/matrix.js

```diff
@@ -0,0 +1,171 @@
+'use strict'; // eslint-disable-line strict
+
+/**
+* Running tests with multiple configurations using varying parameters
+* @constructor
+*/
+class TestMatrix {
+    /**
+    * Builds a new testing Matrix
+    *
+    * @param {Object.<string, string|string[]>} params - Argument of the matrix
+    *
+    * @returns {TestMatrix} new testing Matrix
+    */
+    constructor(params) {
+        /**
+        * Create deep copy for forbidden the user to modify the object outside
+        * the class
+        **/
+        this.params = Object.assign({}, params);
+        this.paramsOnString = JSON.stringify(this.params);
+        this.callback = null;
+        this.elementsToSpecialize = [];
+        this.listOfException = [];
+    }
+
+    /**
+    * Callback used for a generated matrix
+    *
+    * @callback GenerateMatrixCallback
+    * @param {TestMatrix} matrix - The following generate matrix.
+    *
+    * @return {undefined}
+    */
+
+    /**
+    * Generate testing matrix, call execute method to run callback
+    *
+    * @param {string[]} elementsToSpecialize - key element we want to
+    *                    specialize on the matrix
+    * @param {GenerateMatrixCallback} callback - callback function
+    *                    who want to be call on execute method
+    *                    for the current generate matrix
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    generate(elementsToSpecialize, callback) {
+        this.elementsToSpecialize = elementsToSpecialize;
+        this.callback = callback;
+        return this;
+    }
+
+    /**
+    * Method to prevent an exception during the tests.
+    *
+    * @param {Object.<string, string[]>} exception - The keys of the hash table
+    is the element we want to do an exception, the value
+    *                      is the value we want to do an exception for the
+    *                      specific key.
+    *                      All key of the object must be validate with at least
+    *                      one value for execute an exception.
+    *                      ex: for a hashmap with 'auth' as key and 'v4' as
+    *                      value, if we found a matrix with this following
+    *                      paramater, it will call the special callback
+    * @argument {GenerateMatrixCallback} callbackException - Callback who the
+    *                       execute method will call when
+    *                       this exception is found.
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    if(exception, callbackException) {
+        const pair = {
+            key: exception,
+            callback: callbackException,
+        };
+        this.listOfException.push(pair);
+
+        return this;
+    }
+
+    /**
+    * Return the parameters of the matrix into a Json format
+    *
+    * @returns {string} JSON format of parameters
+    */
+    toString() {
```

Comment:
`toString()` is a method JS provides for all objects, I would recommend naming it something else like `serialize`